### PR TITLE
TYP: Replace wildcardimports in toplevelinit as precursor for reshape,stata,io PRs #25936 #25940 #25939

### PR DESCRIPTION
**Pull request by ryankarlos**
- [x] xref #25932, #25933, #25934
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by ryankarlos** [comment]
@WillAyd Since im working on PRs #25936 #25940 #25939  which all depend on this top level import - thought it would make sense to replace wildcard imports for stata, reshape and io modules at the same time.

**Comment by WillAyd** [comment]
I think you can still import from the apis exposed by each package but just need to list out the object that you are pulling in. Does that no work with Mypy?

**Comment by ryankarlos** [comment]
So the only error that mypy throws is this - 
`pandas/__init__.py:70: error: Module 'pandas._libs' has no attribute 'join'`

Theres a lot of other modules in those apis which do not particularly affect the reshape, stata PRs that i am working on but Ive I've included them in in case they need to be imported via top level from other modules in the future.

File: pandas/__init__.py [hunk 1]
```diff
@@ -42,14 +42,37 @@
...
+    concat, lreshape, melt, wide_to_long, merge, merge_asof,
+    merge_ordered, crosstab, pivot, pivot_table, get_dummies,
+    cut, qcut)
+
+from pandas.core.computation.api import eval
+
+from pandas.core.sparse.api import (
+    SparseArray, SparseDataFrame, SparseSeries, SparseDtype)
+
+from pandas.core.internals import BlockManager, _safe_reshape, make_block
```
Comment (jreback) [inline pandas/__init__.py]: these are internal and should not be imported

File: pandas/__init__.py [hunk 2]
```diff
@@ -42,14 +42,37 @@
...
+from pandas.core.arrays import DatetimeArray, IntervalArray, PeriodArray
+from pandas.core.arrays.sparse import BlockIndex, IntIndex
+
+from pandas._libs import join
```
Comment (jreback) [inline pandas/__init__.py]: this is internal

File: pandas/__init__.py [hunk 3]
```diff
@@ -42,14 +42,37 @@
...
 
 from pandas.util._print_versions import show_versions
-from pandas.io.api import *
```
Comment (jreback) [inline pandas/__init__.py]: you need to do this one as well

**Review by jreback — CHANGES_REQUESTED** [review]
this must be *very* precisely done, IOW importing names explicity as we have tests which exactly guaranteee these imports are no more and no less what is actually imported

**Comment by ryankarlos** [comment]
@jreback Thanks for the review, i've pushed latest changes - hopefully they pass now as I think i have the exact number of imports in.

**Review by WillAyd — CHANGES_REQUESTED** [review]
Looks like there is still one wildcard import for NumPy compat - can you take a look at that here?

File: pandas/__init__.py [hunk 4]
```diff
@@ -42,14 +42,42 @@
 # let init-time option registration happen
 import pandas.core.config_init
 
-from pandas.core.api import *
-from pandas.core.sparse.api import *
-from pandas.tseries.api import *
-from pandas.core.computation.api import *
-from pandas.core.reshape.api import *
+from pandas.core.api import (
```
Comment (jreback) [inline pandas/__init__.py]: can you organize these by section, e.g. something like

```
from pandas.core.api import (
    # dtype
    Int8Dtype.........

   # indexes
    Int64Index.....

  # tseries
   NaT, Period.....

  # conversion
   to_numeric, to_datetime...

  # missing
   isna, isnul.....

  # misc
  np, TimeGrouper....

File: pandas/__init__.py [hunk 5]
```diff
@@ -42,14 +42,42 @@
...
+from pandas.core.reshape.api import (
+    concat, lreshape, melt, wide_to_long, merge, merge_asof,
+    merge_ordered, crosstab, pivot, pivot_table, get_dummies,
+    cut, qcut)
 
 from pandas.util._print_versions import show_versions
-from pandas.io.api import *
+
+from pandas.io.api import (
+    read_clipboard, ExcelFile, ExcelWriter, read_excel,
```
Comment (jreback) [inline pandas/__init__.py]: can you sort these

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by ryankarlos** [comment]
@jreback @WillAyd pushed latest changes through

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ryankarlos

**Comment by jreback** [comment]
if you’d merge master and update any typing PRs can get them in

**Comment by ryankarlos** [comment]
Thanks @jreback , pushed latest updates for dependant PRs #25936 #25940 #25939

**Pull request merged**